packages/solarwinds-apm/src/sampling/grpc.ts

```diff
@@ -0,0 +1,375 @@
+  readonly ready: Promise<void>
+  #ready!: () => void
+
+  constructor(config: SwConfiguration, logger: DiagLogger) {
+    super(config, logger)
+
+    this.#key = `${config.token}:${config.serviceName}`
+
+    // convert the collector string into a valid full URL
+    let collector = config.collector!
+    if (!/:{0-9}+$/.test(collector)) {
+      collector = `${collector}:443`
+    }
+    if (!/^https?:/.test(collector)) {
+      collector = `https://${collector}`
+    }
+
+    try {
+      this.#address = new URL(collector)
+
+      // on Alpine the grpc.Client constructor will hang forever if the hostname can't resolve
+      // to avoid this we try to resolve before actually instantiating it
+      const resolve = Promise.any([
+        dns.resolve4(this.#address.hostname),
+        dns.resolve6(this.#address.hostname),
+      ])
+
+      // create a temporary client that checks the hostname can be resolved
+      // before replacing itself with the real thing
+      this.#client = {
+        getSettings: (request, response) =>
+          resolve
+            .then(() => {
+              const cred = config.certificate
+                ? credentials.createSsl(Buffer.from(config.certificate))
+                : credentials.createSsl()
+
+              // at this point we know the hostname resolves so
+              // we can replace the temporary client
+              this.#client = new GrpcCollectorClient(this.#address.host, cred)
+              return this.#client.getSettings(request, response)
+            })
+            .catch((cause: unknown) =>
+              // if the hostname doesn't resolve we just always return an error
+              Promise.reject(
+                new Error(`Invalid collector (${config.collector!})`, {
+                  cause,
+                }),
+              ),
+            ),
+      }
+    } catch (cause) {
+      // this should only happen if the collector setting is set
+      // to complete gibberish
+      this.#address = new URL("https://collector.invalid:443")
+      this.#client = {
+        getSettings: () =>
+          Promise.reject(
+            new Error(`Invalid collector (${config.collector!})`, {
+              cause,
+            }),
+          ),
+      }
+    }
+
+    this.ready = new Promise((resolve) => (this.#ready = resolve))
+
+    setImmediate(() => {
+      this.#loop()
+    }).unref()
+  }
+
+  override toString(): string {
+    return `gRPC Sampler (${this.#address.host})`
+  }
+
+  /** Logs a de-duplicated warning */
+  #warn(message: string, ...args: unknown[]) {
+    if (message === this.#lastWarningMessage) {
+      return
+    }
+
+    this.logger.warn(message, ...args)
+    this.#lastWarningMessage = message
+  }
+
+  #loop(retryTimeout = RETRY_MIN_TIMEOUT) {
+    const retry = () => {
+      this.#ready()
+
+      this.logger.debug(`retrying in ${(retryTimeout / 1000).toFixed(1)}s`)
+      const nextRetryTimeout = Math.min(
+        retryTimeout * MULTIPLIER,
```

Comment:
Is it possible to test the exponential backoff algorithm for the connection failure?

It could be a manual test and put the test result here. Also, Java and C-lib had different implementation, please check it out as I haven't touch this part yet. see [here](https://github.com/solarwinds-cloud/trace/issues/143#issuecomment-265825630) 

Here are some information FYI.
https://github.com/solarwinds-cloud/oboe/issues/506#issuecomment-267415858
https://github.com/solarwinds-cloud/trace/issues/143#issuecomment-265825630
https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html

Reply:
I extracted the backoff logic to its own file to make it easier to test, and added a small test of its integration with the sampler. I also tested it manually and it works as expected.

```
DEBUG   | retrying in 0.5s
DEBUG   | retrying in 0.8s
DEBUG   | retrying in 1.1s
DEBUG   | retrying in 1.7s
DEBUG   | retrying in 2.5s
DEBUG   | retrying in 3.8s
DEBUG   | retrying in 5.7s
DEBUG   | retrying in 8.5s
DEBUG   | retrying in 12.8s
DEBUG   | retrying in 19.2s
DEBUG   | retrying in 28.8s
DEBUG   | retrying in 43.2s
DEBUG   | retrying in 60.0s
DEBUG   | retrying in 60.0s
```

Reply:
Also changed to backoff constants to the same as liboboe for consistency